src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -1765,6 +1776,95 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             suggested_limit, self.tcx.crate_name,
         ));
     }
+
+    fn suggest_await_before_try(
+        &self,
+        err: &mut DiagnosticBuilder<'_>,
+        obligation: &PredicateObligation<'tcx>,
+        trait_ref: &ty::Binder<ty::TraitRef<'tcx>>,
+        span: Span,
+    ) {
+        debug!(
+            "suggest_await_befor_try: obligation={:?}, span={:?}, trait_ref={:?}",
+            obligation, span, trait_ref
+        );
+        let body_hir_id = obligation.cause.body_id;
+        let item_id = self.tcx.hir().get_parent_node(body_hir_id);
+
+        let mut is_future = false;
+        if let ty::Opaque(def_id, substs) = trait_ref.self_ty().kind {
```

Comment:
it may also be a non-opaque future. Instead of searching for the future trait, you could also try to have the trait system check `trait_ref.self_ty(): Future` by using a scheme similar to clippy's `implements_trait`: https://github.com/rust-lang/rust/blob/bad3bf622bded50a97c0a54e29350eada2a3a169/src/tools/clippy/clippy_lints/src/utils/mod.rs#L323

Ideally, if there is a lot of code overlap between the clippy version and what you need, move most of this function onto `TyCtxt` and call it from both here and from clippy.